### DISCUSSION TITLE
Building a dist and packaging

### DISCUSSION THREAD

**Discussion by crosser** [svinota/pyroute2#1053] [Q&A]
I found that since a while ago (around version 0.7 perhaps?) `setup.py` files were removed. That broke packaging via setuptools, that my in-house debianization was using.

I would like to know if people build .deb packages of this library these days (except Debian, their debianization relies on openstack infrastructure and therefore un-reusable)?

By the way there is a snag in the "recommended" build procedure, the script that looks for python returns nothing if `venv` is not installed, and then `make` fails with a confusing error.

In any case, I suspect that building with make/pip won't work in isolated environment, because I noticed that pip was downloading hoards of packages when I ran it by hand? Is that correct?

Anyway, my main question is: what is the recommended way to build a distribution-ready package in isolated CI environment?

**Comment by svinota** [comment]
Yep, unlike the runtime that doesn't require extra packages, the build process requires the whole load of packages like setuptools, wheel, build, twine etc. And I was sure that removal of `setup.py` will not break the old-style setuptools build, but looks like I was wrong.

What restrictions do you have on these environments? Is there any, say, docker image or something like that, so I could run and try to adapt the build process for the environment?

Action items for me (feel free to propose more):

- [ ] Get / create such environment
- [ ] Make setup running there w/o network connectivity
- [ ] Analyze the routine and include critical parts into the CI
- [ ] Probably — establish a routine to introduce major changes getting feedback from the community before merging into the master branch

**Reply by crosser** [reply]
I don't have a docker image, and as far as I know, our CI (that is in the hands of a different team) does not currently use docker.

Still, the requirements are quite simple and basic: to make our CI build my package I have to provide three things:

1. list if build dependencies (in the form of debian packages),
2. the source tree of the package (populated directory or tarball will do)
3. the commands that that would build and install the package (run under `fakeroot`, as far as I know).

I have the list of dependencies for the previous version, and I have no problem to update it myself for the new version. I have no problem building the populated directory, for example by recursive copy of freshly checked out version, excluding `.git` subdir.

The problem is the build command. Debian package builder uses setuptools by default, so just telling it "this is a python3 package" did the right thing. Debian builder called `python3 setup.py <relevant command>`, that installed everything in the fake root, and packaged this fake root as `.deb` package(s). And this process is performed in the environment that does not have Internet access, which is rather an obvious requirement from the standpoint of reproducibility.

This last part does not work now because Debian builder cannot invoke setuptools. It _can_ be told to run some other commands. As the lowest common denominator, I can tell it to just run `make` with appropriate target. I did not try it, because I am relatively certain that it won't work, I saw that when run by hand, it downloads things from the Internet.

**Reply by svinota** [reply]
@crosser could you please take a look at this PR: https://github.com/svinota/pyroute2/pull/1055 , if it will work for you
